shared/src/ledger/ibc.rs

```diff
@@ -1,36 +1,419 @@
+use ibc::ics02_client::client_consensus::AnyConsensusState;
+use ibc::ics02_client::client_state::AnyClientState;
+use ibc::ics02_client::client_type::ClientType;
+use ibc::ics02_client::context::ClientReader;
+use ibc::ics02_client::height::Height;
+use ibc::ics24_host::identifier::ClientId;
+use ibc::ics24_host::Path;
+use tendermint_proto::Protobuf;
+use thiserror::Error;
+
+use crate::ledger::native_vp::{self, Ctx, NativeVp};
 use crate::ledger::storage::{self, Storage, StorageHasher};
-use crate::ledger::vp_env::Result;
 use crate::types::address::{Address, InternalAddress};
-use crate::types::storage::Key;
+use crate::types::storage::{Key, KeySeg};
+
+#[allow(missing_docs)]
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Native VP error: {0}")]
+    NativeVpError(native_vp::Error),
+    #[error("Key error: {0}")]
+    KeyError(String),
+}
+
+/// IBC functions result
+pub type Result<T> = std::result::Result<T, Error>;
 
 /// IBC VP
-pub struct Ibc;
+pub struct Ibc<'a, DB, H>
+where
+    DB: storage::DB + for<'iter> storage::DBIter<'iter>,
+    H: StorageHasher,
+{
+    /// Context to interact with the host structures.
+    pub ctx: Ctx<'a, DB, H>,
+}
+
+impl<'a, DB, H> NativeVp for Ibc<'a, DB, H>
+where
+    DB: 'static + storage::DB + for<'iter> storage::DBIter<'iter>,
+    H: 'static + StorageHasher,
+{
+    type Error = Error;
 
-impl NativeVp for Ibc {
     const ADDR: InternalAddress = InternalAddress::Ibc;
 
-    fn init_genesis_storage<DB, H>(_storage: &mut Storage<DB, H>)
+    fn init_genesis_storage<D, SH>(storage: &mut Storage<D, SH>)
     where
-        DB: storage::DB + for<'iter> storage::DBIter<'iter>,
-        H: StorageHasher,
+        D: storage::DB + for<'iter> storage::DBIter<'iter>,
+        SH: StorageHasher,
     {
+        // the client counter
+        let path = "clients/counter".to_owned();
+        let key = Key::ibc_key(path)
+            .expect("Creating a key for a client counter failed");
+        let value = crate::ledger::storage::types::encode(&0);
+        storage
+            .write(&key, value)
+            .expect("Unable to write the initial client counter");
     }
 
-    fn validate_tx<DB, H>(
-        _ctx: &mut Ctx<DB, H>,
+    fn validate_tx(
+        &self,
         _tx_data: &[u8],
-        _keys_changed: &HashSet<Key>,
+        keys_changed: &HashSet<Key>,
         _verifiers: &HashSet<Address>,
-    ) -> Result<bool>
-    where
-        DB: storage::DB + for<'iter> storage::DBIter<'iter>,
-        H: StorageHasher,
-    {
+    ) -> Result<bool> {
+        let mut clients = HashSet::new();
+
+        for key in keys_changed {
+            if !key.is_ibc_key() {
+                continue;
+            }
+
+            let accepted = match Self::get_ibc_prefix(key) {
+                IbcPrefix::Client => {
+                    let client_id = Self::get_client_id(key)?;
+                    if !clients.insert(client_id.clone()) {
+                        // this client has been checked
```

Comment:
why would we be iterating over it twice, then? what if a different (not necessarily valid) value is written?

Reply:
Thank you for your comment.
It avoids the repeated validations because `clients/{identifier}/clientType`, `clients/{identifier}/clientState` and `clients/{identifier}/consensusStates/{height}` are checked at once.
However, it looks better to check the suffix and validate each state.

Reply:
ah I see, never mind